witnesslint.py

```diff
@@ -0,0 +1,1128 @@
+        if self.programfile is None:
+            self.log(LOGLEVELS["warning"], "Programfile has not been specified")
+        if self.programhash is None:
+            self.log(LOGLEVELS["warning"], "Programhash has not been specified")
+        if self.architecture is None:
+            self.log(LOGLEVELS["warning"], "Architecture has not been specified")
+        if self.creationtime is None:
+            self.log(LOGLEVELS["warning"], "Creationtime has not been specified")
+        if self.entry_node is None and len(self.node_ids) > 0:
+            self.log(LOGLEVELS["warning"], "No entry node has been specified")
+        if self.witness_type == "correctness_witness":
+            for key in self.violation_witness_only:
+                self.log(
+                    LOGLEVELS["warning"],
+                    "Key '%s' is not allowed in correctness witness",
+                    key,
+                )
+        elif self.witness_type == "violation_witness":
+            for key in self.correctness_witness_only:
+                self.log(
+                    LOGLEVELS["warning"],
+                    "Key '%s' is not allowed in violation witness",
+                    key,
+                )
+        for node_id in self.check_existence_later:
+            if node_id not in self.node_ids:
+                self.log(LOGLEVELS["warning"], "Node %s has not been declared", node_id)
+        if self.check_callstack:
+            self.check_function_stack(
+                collections.OrderedDict(sorted(self.transitions.items())),
+                self.entry_node,
+            )
+        if self.check_program and self.program_info is not None:
+            for check in self.check_later:
+                check()
+
+    def lint(self):
+        """
+        Splits the witness into manageable chunks and triggers or performs
+        checks for the resulting elements. Also stores some information to be
+        able to trigger checks for the witness as a whole.
+        """
+        try:
+            saw_graph = False
+            saw_graphml = False
+            element_stack = list()
+            for (event, elem) in ET.iterparse(self.witness, events=("start", "end")):
+                if event == "start":
+                    element_stack.append(elem)
+                else:
+                    element_stack.pop()
+                    _, _, tag = elem.tag.rpartition("}")
+                    if tag == "data":
+                        # Will be handled later
+                        pass
+                    elif tag == "default":
+                        # Will be handled later
+                        pass
+                    elif tag == "key":
+                        self.handle_key(elem)
+                        element_stack[-1].remove(elem)
+                    elif tag == "node":
+                        self.handle_node(elem)
+                        element_stack[-1].remove(elem)
+                    elif tag == "edge":
```

Comment:
Please use constants for these kinds of keywords, e.g., `NODE = "node"`, and similarly for "key", "edge", etc.

Reply:
Done.